### PR TITLE
Support for stopping container

### PR DESCRIPTION
**Pull request by YaoZengzeng**
Just directly use kill SIGKILL to stop the container for temporary.
And modify "stop command of hyperctl" like "kill" which can stop pod or container.

### CODE REVIEW

File: daemon/stop.go [hunk 1]
```diff
@@ -89,3 +90,21 @@ func (daemon *Daemon) StopPodWithinLock(pod *Pod) (int, string, error) {
...
+	containerId := p.status.Containers[idx].Id
+	glog.V(1).Infof("found container %s to stop", containerId)
+
+	err := p.vm.KillContainer(containerId, syscall.SIGTERM)
+	if err == nil {
+		return nil
+	}
+	err = p.vm.KillContainer(containerId, syscall.SIGKILL)
+	return err
+	//TODO: update container status of the pod
```
Comment (feiskyer) [inline daemon/stop.go]: @laijs  Any updates on pod/container state machine?

File: types/types.proto [hunk 2]
```diff
@@ -541,6 +549,8 @@ service PublicAPI {
     // TODO: ContainerCommit commits the changes of the specified container
     // TODO: ContainerSignal sends a singla to specified container
     // TODO: ContainerLabels updates labels of the specified container
+    // ContainerStop implements POST /container/stop
```
Comment (feiskyer) [inline types/types.proto]: ContainerStop stops the specified container

Comment (feiskyer) [inline types/types.proto]: `/container/stop` restful api and client are also required.

File: daemon/stop.go (hunk 1 above)
Comment (YaoZengzeng) [inline daemon/stop.go]: ref to runv pr #251

File: daemon/stop.go [hunk 3]
```diff
@@ -89,3 +90,21 @@ func (daemon *Daemon) StopPodWithinLock(pod *Pod) (int, string, error) {
...
+		return fmt.Errorf("can not found container %s", container)
+	}
+
+	containerId := p.status.Containers[idx].Id
+	glog.V(1).Infof("found container %s to stop", containerId)
+
+	err := p.vm.KillContainer(containerId, syscall.SIGTERM)
+	if err == nil {
+		return nil
+	}
```
Comment (laijs) [inline daemon/stop.go]: p.vm.KillContainer(containerId, syscall.SIGTERM) hardly fail. but the process may continue to run.
please wait/check the container's status(we haven't had such api) in several seconds and then kill SIGKILL if needed.
or directly kill SIGKILL.

File: daemon/stop.go [hunk 4]
```diff
@@ -89,3 +90,17 @@ func (daemon *Daemon) StopPodWithinLock(pod *Pod) (int, string, error) {
...
+	if !ok {
+		return fmt.Errorf("can not found container %s", container)
+	}
+
+	containerId := p.status.Containers[idx].Id
+	glog.V(1).Infof("found container %s to stop", containerId)
+
+	err := p.vm.KillContainer(containerId, syscall.SIGKILL)
+	return err
+	//TODO: update container status of the pod
```
Comment (feiskyer) [inline daemon/stop.go]: Could you update the status of container and keep pod running even container is stopped?

Comment (YaoZengzeng) [inline daemon/stop.go]: Yes,I'll do it in next pr.
Just call  "SetOneContainerStatus" in runv is fine.

File: daemon/stop.go [hunk 5]
```diff
@@ -89,3 +90,17 @@ func (daemon *Daemon) StopPodWithinLock(pod *Pod) (int, string, error) {
...
+func (daemon *Daemon) StopContainer(container string) error {
+	p, idx, ok := daemon.PodList.GetByContainerIdOrName(container)
+	if !ok {
+		return fmt.Errorf("can not found container %s", container)
+	}
+
+	containerId := p.status.Containers[idx].Id
+	glog.V(1).Infof("found container %s to stop", containerId)
+
+	err := p.vm.KillContainer(containerId, syscall.SIGKILL)
```
Comment (gao-feng) [inline daemon/stop.go]: you should check if p.vm is nil (pod is not started yet)

File: client/stop.go [hunk 6]
```diff
@@ -12,10 +12,11 @@ import (
 func (cli *HyperClient) HyperCmdStop(args ...string) error {
 
 	var opts struct {
+		Pod  bool `short:"p" long:"pod" default:"false" default-mask:"-" description:"stop a pod"`
```
Comment (gao-feng) [inline client/stop.go]: Should stop pod by default.

File: daemon/stop.go [hunk 7]
```diff
@@ -89,3 +90,21 @@ func (daemon *Daemon) StopPodWithinLock(pod *Pod) (int, string, error) {
 
 	return vmResponse.Code, vmResponse.Cause, nil
 }
+
+func (daemon *Daemon) StopContainer(container string) error {
+	pod, idx, ok := daemon.PodList.GetByContainerIdOrName(container)
+	if !ok {
+		return fmt.Errorf("can not found container %s", container)
+	}
+
```
Comment (YaoZengzeng) [inline daemon/stop.go]: Transition Lock may need there.But I'll do it in the next pr which update  container status of the pod.

**Comment by feiskyer** [comment]
lgtm

**Pull request merged**